### PR TITLE
Set io implementation as provided for smb and parquet

### LINKED ISSUE #4327: SMB module is pulling all storage implementations
Depending on `scio-smb` pulls transitively all storage implementation dependencies for:
- parquet
- json
- avro
- tensorflow

TensorFlow dependencies alone are ~200Mb.

Users should only have the desired storage implementation in their classpath

SMB should either:
- be available in the storage implementation themselves
- mark storage impl dependencies as provided
- have specific impl modules like

### LINKED ISSUE #4788: Make tensorflow dependencies provided in scio-parquet?
The Tensorflow jars are large and most scio-parquet users are only using Avro or Case Class functionality.

### PR DESCRIPTION
**Pull request by RustedBones**
Fix #4788
Fix #4327

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1047,11 +1047,7 @@ lazy val `scio-parquet`: Project = project
       "com.spotify" %% "magnolify-parquet" % magnolifyVersion,
       "com.twitter" %% "chill" % chillVersion,
       "me.lyh" % "parquet-tensorflow" % parquetExtraVersion,
-      "me.lyh" %% "parquet-avro" % parquetExtraVersion excludeAll (
-        // parquet-avro depends on avro 1.10.x
-        "org.apache.avro" % "avro",
-        "org.apache.avro" % "avro-compiler"
```
Comment (RustedBones) [inline build.sbt:1053]: Libs are marked as `provided` in `me.lyh:parquet-avro`

File: build.sbt [hunk 2]
```diff
@@ -1344,25 +1341,25 @@ lazy val `scio-smb`: Project = project
...
       "org.slf4j" % "slf4j-api" % slf4jVersion,
       // provided
       "com.github.ben-manes.caffeine" % "caffeine" % caffeineVersion % Provided,
+      "org.apache.avro" % "avro" % avroVersion % Provided,
+      "org.apache.hadoop" % "hadoop-common" % hadoopVersion % Provided,
+      "org.apache.parquet" % "parquet-avro" % parquetVersion % Provided excludeAll ("org.apache.avro" % "avro"),
+      "org.apache.parquet" % "parquet-column" % parquetVersion % Provided,
+      "org.apache.parquet" % "parquet-common" % parquetVersion % Provided,
+      "org.apache.parquet" % "parquet-hadoop" % parquetVersion % Provided,
+      "org.tensorflow" % "tensorflow-core-api" % tensorFlowVersion % Provided,
```
Comment (RustedBones) [inline build.sbt:1362]: For smb, I think it is safe to assume pple will also have `scio-avro` `scio-parquet` or `scio-tensorflow` as depenedecy. Do not pull everything here

File: build.sbt [hunk 3]
```diff
@@ -1068,7 +1064,8 @@ lazy val `scio-parquet`: Project = project
       "org.apache.parquet" % "parquet-hadoop" % parquetVersion,
       "org.scala-lang.modules" %% "scala-collection-compat" % scalaCollectionCompatVersion,
       "org.slf4j" % "slf4j-api" % slf4jVersion,
-      "org.tensorflow" % "tensorflow-core-api" % tensorFlowVersion,
+      // provided
+      "org.tensorflow" % "tensorflow-core-api" % tensorFlowVersion % Provided,
```
Comment (clairemcginty) [inline build.sbt:1068]: 👍

File: build.sbt (hunk 2 above)
Comment (clairemcginty) [reply build.sbt:1362]: hm I do think it's fairly common to have scio-smb but not scio-parquet in your dependencies; I think this change makes sense, but let's definitely call it out in the migration guide.

**Review by clairemcginty — APPROVED** [review]

**Review by clairemcginty — APPROVED** [review]

File: build.sbt (hunk 2 above)
Comment (shnapz) [reply build.sbt:1362]: In theory we can wrap `ClassNotFoundException` with more meaningful instructions. Even with good docs, there will be confused users writing to support channels

Comment (RustedBones) [reply build.sbt:1363]: I added some checks in the static class/object initialization of the various SMB impl

**Review by kellen — APPROVED** [review]

**Pull request merged**